# Lab 2 — Regression (California Housing)

**Objective:** Predict median house value for California census block groups.

**Dataset:** California Housing — 20,640 samples, 8 features (MedInc, HouseAge, AveRooms, etc.), target = median house value (in 100,000 USD).

## Step 1: Imports and load data

Import the required libraries and load the California Housing dataset. Store features in `X` and target in `y`.

In [1]:
# Step 1: Imports and load data
from sklearn.datasets import fetch_california_housing
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import numpy as np

# Load California Housing data
data = fetch_california_housing()
X = data.data
y = data.target

print("X shape:", X.shape)
print("y shape:", y.shape)
print("Feature names:", data.feature_names)

X shape: (20640, 8)
y shape: (20640,)
Feature names: ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']


## Step 2: Train–test split

Split `X` and `y` into training and test sets with `test_size=0.3`.

In [2]:
# Step 2: Train–test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

print("Train size:", len(y_train))
print("Test size:", len(y_test))

Train size: 14448
Test size: 6192


## Step 3: Train the model

Create a `LinearRegression` model, fit it on the training data, and predict on the test set.

In [3]:
# Step 3: Train the model
model = LinearRegression()
model.fit(X_train, y_train)
pred = model.predict(X_test)

print("Model training completed")

Model training completed


## Step 4: Evaluate

Print the Mean Squared Error between `y_test` and `pred`.

In [4]:
# Step 4: Evaluate
mse = mean_squared_error(y_test, pred)
rmse = np.sqrt(mse)

print(f"MSE: {mse:.4f}")
print(f"RMSE: {rmse:.4f}")

MSE: 0.5306
RMSE: 0.7284


---
## Lab questions (homework)

Answer in the cell below or in a separate document.

1. **Interpret the MSE.** What does it measure? Is your value "good" or "bad"? (Remember: target is in units of 100,000 USD.) You can also compute RMSE = √MSE.

2. **What features influence price most?** Use `data.feature_names` and `model.coef_` to list each feature and its coefficient. Which has the largest positive effect? Largest negative? Briefly interpret.

*Your answers here:*

1. MSE (Mean Squared Error) measures the average squared difference between predicted and actual values. With MSE ≈ 0.53 and RMSE ≈ 0.73 (in units of 100,000 USD), this means:

On average, our predictions are off by about $73,000 from the actual house values

Given that house values in this dataset typically range from about $50,000 to $500,000, an error of $73,000 is moderately high

This suggests the linear model has decent predictive power but significant room for improvement

2. From the coefficients:

Largest positive effect: AveBedrms (0.6242) - More bedrooms per household increases house value, with each additional bedroom per household associated with a $62,420 increase in house value

Largest negative effect: Longitude (-0.4342) - Moving eastward decreases house value, reflecting California's geography where coastal (western) areas are more expensive